### PR TITLE
Support Record Expression

### PR DESCRIPTION
**Pull request by yoshihiro503**
for #54 

Support following expressions.
* record creation:  `#user{name = "Taro", admin = true}`
* record field access: `U#user.name`
* record field index number: `#user.name`
* record update: `U#user{admin = true}`

### CODE REVIEW

File: lib/abstract_format.ml [hunk 1]
```diff
@@ -596,6 +603,56 @@ and expr_of_sf sf : (expr_t, err_t) Result.t =
...
+  (* a record creation : #user{name = "Taro", admin = true} *)
+  | Sf.Tuple (4, [Sf.Atom "record";
+                  Sf.Integer line;
+                  Sf.Atom name;
+                  Sf.List sf_record_fields]) ->
+     let field_of_sf sf =
+       begin match record_field_of_sf sf with
+       | Ok (RecordField {line; field_name; ty=None; default_expr=Some e}) ->
+          (line, field_name, e) |> return
+       | Ok _ -> failwith "cannot reach here"
```
Comment (amutake) [inline lib/abstract_format.ml]: Could you use [`Err.Invalid_input`](https://github.com/yutopp/obeam/blob/d4974b8f2542d88cf14e89f33dc2bdba1bcfcdec/lib/err.ml#L18)?

File: lib/abstract_format.ml [hunk 2]
```diff
@@ -596,6 +603,56 @@ and expr_of_sf sf : (expr_t, err_t) Result.t =
...
+  | Sf.Tuple (5, [Sf.Atom "record";
+                  Sf.Integer line;
+                  sf_expr;
+                  Sf.Atom name;
+                  Sf.List sf_update_fields]) ->
+     let update_field_of_sf sf =
+       begin match record_field_of_sf sf with
+       | Ok (RecordField {line; field_name; ty=None; default_expr=Some e}) ->
+          (line, field_name, e) |> return
+       | Ok _ -> failwith "cannot reach here"
```
Comment (amutake) [inline lib/abstract_format.ml]: ditto

File: lib/abstract_format.ml [hunk 3]
```diff
@@ -596,6 +603,56 @@ and expr_of_sf sf : (expr_t, err_t) Result.t =
...
+       | Error e -> Error e
+       end
+     in
+     let%bind expr = sf_expr |> expr_of_sf in
```
Comment (amutake) [inline lib/abstract_format.ml]: Could you add `|> track ~loc:[%here]`?

File: lib/abstract_format.ml (hunk 1 above)
Comment (yoshihiro503) [reply lib/abstract_format.ml]: fixed.

File: lib/abstract_format.ml (hunk 2 above)
Comment (yoshihiro503) [reply lib/abstract_format.ml]: fixed.

File: lib/abstract_format.ml (hunk 3 above)
Comment (yoshihiro503) [reply lib/abstract_format.ml]: thanks

File: lib/abstract_format.ml [hunk 4]
```diff
@@ -596,6 +603,58 @@ and expr_of_sf sf : (expr_t, err_t) Result.t =
...
+  | Sf.Tuple (4, [Sf.Atom "record";
+                  Sf.Integer line;
+                  Sf.Atom name;
+                  Sf.List sf_record_fields]) ->
+     let field_of_sf sf =
+       begin match record_field_of_sf sf with
+       | Ok (RecordField {line; field_name; ty=None; default_expr=Some e}) ->
+          (line, field_name, e) |> return
+       | Ok _ ->
+          Err.create ~loc:[%here] (Err.Invalid_input ("cannot reach here", sf)) |> Result.fail
```
Comment (amutake) [inline lib/abstract_format.ml]: Could you make error message more detailed? e.g., "record field must have field name and expression" 🙏

File: lib/abstract_format.ml [hunk 5]
```diff
@@ -596,6 +603,58 @@ and expr_of_sf sf : (expr_t, err_t) Result.t =
...
+                  Sf.Integer line;
+                  sf_expr;
+                  Sf.Atom name;
+                  Sf.List sf_update_fields]) ->
+     let update_field_of_sf sf =
+       begin match record_field_of_sf sf with
+       | Ok (RecordField {line; field_name; ty=None; default_expr=Some e}) ->
+          (line, field_name, e) |> return
+       | Ok _ ->
+          Err.create ~loc:[%here] (Err.Invalid_input ("cannot reach here", sf)) |> Result.fail
```
Comment (amutake) [inline lib/abstract_format.ml]: ditto 🙏

File: lib/abstract_format.ml (hunk 4 above)
Comment (yoshihiro503) [reply lib/abstract_format.ml]: I see 🙏

File: lib/abstract_format.ml (hunk 5 above)
Comment (yoshihiro503) [reply lib/abstract_format.ml]: I see 🙏

**Pull request merged**